pkg/registries/registry.go

```diff
@@ -0,0 +1,166 @@
+package registries
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	logging "github.com/op/go-logging"
+	"github.com/openshift/ansible-service-broker/pkg/apb"
+	"github.com/openshift/ansible-service-broker/pkg/registries/adapters"
+)
+
+var regex = regexp.MustCompile(`[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*`)
+
+// Config - Configuration for the registry
+type Config struct {
+	URL  string
+	User string
+	Pass string
+	Org  string
+	Type string
+	Name string
+	// Fail will tell the registry that it is ok to fail the bootstrap if
+	// just this registry has failed.
+	Fail      bool     `yaml:"fail_on_error"`
+	WhiteList []string `yaml:"white_list"`
+	BlackList []string `yaml:"black_list"`
+}
+
+// Validate - makes sure the registry config is valid.
+func (c Config) Validate() bool {
+	if c.Name == "" {
+		return false
+	}
+	m := regex.FindString(c.Name)
+	return m == c.Name
+}
+
+// Registry - manages an adapter to retrieve and manage images to specs.
+type Registry struct {
+	config  Config
+	adapter adapters.Adapter
+	log     *logging.Logger
+	filter  Filter
+}
+
+// LoadSpecs - Load the specs for the registry.
+func (r Registry) LoadSpecs() ([]*apb.Spec, int, error) {
+	imageNames, err := r.adapter.GetImageNames()
+	if err != nil {
+		r.log.Errorf("unable to retrieve image names for registry %v - %v",
+			r.config.Name, err)
+		return []*apb.Spec{}, 0, err
+	}
+	validNames, filteredNames := r.filter.Run(imageNames)
+
+	r.log.Debug("Filter applied against registry: %s", r.config.Name)
+
+	if len(validNames) != 0 {
+		r.log.Debugf("Passing APBs:", r.config.Name)
+		for _, name := range validNames {
+			r.log.Debugf("-> %s", name)
+		}
+	}
+
+	if len(filteredNames) != 0 {
+		r.log.Info("Filtered APBs:")
```

Comment:
I'm realizing my initial assumption was wrong, but I still think the set of image names that hits the filter should *ideally* be only APBs. I originally thought the adapter was pushing the query serverside (GET /<whatever>?name=*-apb) or something to that effect, but that's actually not how its *ever* worked. It has always fetched every name under an org, and we filtered on our side based on *-apb. It was always a limitation of their API, from the work we did back in CAP days.

My thoughts on an ideal scenario:
1) Adapters can query a service and get back *only APBs*
2) Registry is able to validate this set and confirm that the names are indeed apbs
3) Vetted set of names hits filter
4) By this point, we're confident we have a set of images we know we care about, and we load the specs. (Possibly validate additional labels like apb.spec version).

Obviously we need to make some concessions because:
1) The service a particular adapter is talking to for fetching APB names may or may not support server-side filtering. In Dockerhub's case, it definitely does not.
2) Since we're strictly dealing with names, we don't have a lot of options for determining whether or not an image is actually an APB. Until now, it has always been `-apb` suffix.

I'd propose `Registry.LoadSpecs` to  filter any image names that do not match `-apb$` before they hit `Filter.Run`. It's a filter that applies to all registries, so it belongs at the Registry level and not in the adapters. Potential drawback here is possible apbs will be filtered if they don't follow this naming convention, but that has always been the case until now, and to your point, we should put a card on the backlog to think about how we can accomplish this better.